app/terminal/terminal.tsx

```diff
@@ -72,7 +80,33 @@ export default class TerminalComponent extends React.Component<TerminalProps, Te
   }
 
   handleDownloadClicked() {
-    var element = document.createElement("a");
+    if (this.props.fullLogsFetcher) {
+      this.setState({ isLoadingFullLog: true });
+      const element = document.createElement("a");
+      this.props
+        .fullLogsFetcher()
+        .then((fullLog: string) => {
+          const unstyledLogs = fullLog.replace(ANSI_STYLES_REGEX, "");
+          element.setAttribute("href", "data:text/plain;charset=utf-8," + encodeURIComponent(unstyledLogs));
+          element.setAttribute("download", "build_logs.txt");
+          element.style.display = "none";
+          document.body.appendChild(element);
+          element.click();
+        })
+        .catch((e) => errorService.handleError(e))
+        .finally(() => {
+          this.setState({ isLoadingFullLog: false });
+          try {
+            document.body.removeChild(element);
+          } catch (e) {
+            if (!(e instanceof DOMException) || e.name != "NotFoundError") {
+              errorService.handleError(e);
+            }
+          }
```

Comment:
I don't totally understand this part -- do errors from the download itself only surface when removing the element? (That seems surprising to me -- seems like they should be surfaced regardless of whether or not the element is removed)

It would be nice if the `remove` didn't need special case handling for the `logsFetcher` case, because then the downloading part could be factored into a separate top-level utility function like `downloadFile()` which creates the invisible link, clicks on it, then removes it

Reply:
I didn't know about `element.remove()`, and `removeChild` *could* throw an error if the child doesn't exist, which is to say if any statement before `appendChild` threw an exception. I didn't want to propagate that error. Fortunately, `element.remove()` has no such problems.